ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java

```diff
@@ -245,52 +303,102 @@ public long getBytesRead() {
-    // add type
-    bufWrap.putShort((short) type.ordinal());
-    if (type != Type.Uninitialized) {
-      // add incarnationId
-      bufWrap.putInt(incarnationIdBytes.length);
-      bufWrap.put(incarnationIdBytes);
-      // add sessionId
-      bufWrap.putInt(sessionIdBytes.length);
-      bufWrap.put(sessionIdBytes);
-      // add offset
-      bufWrap.put(offsetBytes);
-      if (type == Type.JournalBased) {
-        bufWrap.put(getInclusive() ? (byte) 1 : (byte) 0);
-      } else if (type == Type.IndexBased) {
+    byte[] buf = null;
+    switch (version) {
+      case VERSION_0:
+        int OFFSET_SIZE = 8;
+        int START_OFFSET_SIZE = 8;
+        byte[] sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        byte[] storeKeyBytes = storeKey != null ? storeKey.toBytes() : ZERO_LENGTH_ARRAY;
+        int size = VERSION_SIZE + SESSION_ID_LENGTH_SIZE + sessionIdBytes.length +
+            OFFSET_SIZE + START_OFFSET_SIZE + storeKeyBytes.length;
+        buf = new byte[size];
+        ByteBuffer bufWrap = ByteBuffer.wrap(buf);
+        // add version
+        bufWrap.putShort(version);
+        // add sessionId
+        bufWrap.putInt(sessionIdBytes.length);
+        bufWrap.put(sessionIdBytes);
+        // add offset for journal based token
+        bufWrap.putLong((type == Type.JournalBased) ? (offset != null ? offset.getOffset() : UNINITIALIZED_OFFSET)
+            : UNINITIALIZED_OFFSET);
+        // add index start offset for Index based token
+        bufWrap.putLong((type == Type.IndexBased) ? (offset != null ? offset.getOffset() : UNINITIALIZED_OFFSET)
+            : UNINITIALIZED_OFFSET);
+        // add storekey
         bufWrap.put(storeKeyBytes);
-      }
+        break;
+      case VERSION_1:
+        byte[] offsetBytes = offset != null ? offset.toBytes() : ZERO_LENGTH_ARRAY;
+        sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        storeKeyBytes = storeKey != null ? storeKey.toBytes() : ZERO_LENGTH_ARRAY;
+        size = VERSION_SIZE + SESSION_ID_LENGTH_SIZE + sessionIdBytes.length + TYPE_SIZE + offsetBytes.length
+            + storeKeyBytes.length;
+        buf = new byte[size];
+        bufWrap = ByteBuffer.wrap(buf);
+        // add version
+        bufWrap.putShort(VERSION_1);
+        // add sessionId
+        bufWrap.putInt(sessionIdBytes.length);
+        bufWrap.put(sessionIdBytes);
+        // add type
+        bufWrap.putShort((short) type.ordinal());
+        // add offset
+        bufWrap.put(offsetBytes);
+        // add StoreKey
+        bufWrap.put(storeKeyBytes);
+        break;
+      case VERSION_2:
+        offsetBytes = offset != null ? offset.toBytes() : ZERO_LENGTH_ARRAY;
+        sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        byte[] incarnationIdBytes = incarnationId != null ? incarnationId.toString().getBytes() : ZERO_LENGTH_ARRAY;
```

Comment:
If this is null, fail to serialize and throw an error. This should never be null in fact, and that should be ensured at the time of construction.